### ISSUE TITLE
Python wrapper doesn't free memory

### ISSUE THREAD

**Issue by petrzel** [JohnLangford/vowpal_wabbit#1268]
from vowpalwabbit import pyvw  # RSS memory 45M
m = pyvw.vw(initial_regressor='foo.model', quiet=True)  # RSS memory 314M
m.finish()  # RSS memory 311M
del(m)  # RSS memory 311M
m = pyvw.vw(initial_regressor='foo.model', quiet=True)  # RSS memory 578M
m.finish()  # RSS memory 580M
del(m)  # RSS memory 580M
m = pyvw.vw(initial_regressor='foo.model', quiet=True)  # RSS memory 843M

https://github.com/JohnLangford/vowpal_wabbit/blob/master/python/pylibvw.cc#L53 - really?

**Comment by JohnLangford** [comment]
@gramhagen @hal3 have you had a chance to think about memory leaks in python issues?

**Comment by gramhagen** [comment]
hmm, I thought python garbage collection would kick in eventually. I think L53 is correct, because the way python continues to reference the model, there's also a line in the sklearn wrapper that calls finish before the model is discarded.
It could be that adding a new method in pylibvw and connecting it to __del__ is a better way to go about it since that is currently commented out and pointing to L53.

```
void my_delete(vw_ptr all)
{ VW::finish(*all);
}

.def("__del__", &my_delete, "deconstruct the VW object")
```

@petrzel if you're able to test something like this that would be awesome, I might have some time later this week otherwise

**Comment by JohnLangford** [comment]
Thanks @gramhagen

**Comment by gramhagen** [comment]
played around with this a bit and not sure there's an easy way to release the memory allocated by python. I implemented the strategy I proposed above but it did not work, memory kept growing despite finishing and deleting objects until it reached a plateau and then python started garbage collection.

Another approach using subprocesses mentioned here seems like a viable alternative. Hopefully that can work.
https://stackoverflow.com/questions/32167386/force-garbage-collection-in-python-to-free-memory
